perf/trace.py

```diff
@@ -0,0 +1,140 @@
+        self.run_type = self.params.get('type', default='upstream')
+        detected_distro = distro.detect()
+
+        if self.run_type == 'distro':
+            # Make sure kernel source repo is configured
+            if detected_distro.name in ['rhel', 'centos']:
+                deps = ['yum-utils']
+                for package in deps:
+                    if not smg.check_installed(package) and not\
+                            smg.install(package):
+                        self.cancel("Fail to install %s\
+                                    required for this test."
+                                    % (package))
+                uname_cmd = "uname -r | cut -f 1-4 -d '.'"
+                self.output = self.run_cmd_out("%s" % uname_cmd).decode(
+                                               "utf-8")
+                kernel_src = ("kernel-%s.src" % self.output)
+
+                download_cmd = ("yumdownloader --assumeyes --verbose\
+                                --source %s --destdir %s" %
+                                (kernel_src, self.workdir))
+                self.run_cmd_out("%s" % download_cmd)
+
+                install_rpm_cmd = ("rpm -i %s/%s.rpm" %
+                                   (self.workdir, kernel_src))
+                self.run_cmd_out("%s" % install_rpm_cmd)
+
+                src_def_path = "/root/rpmbuild/SOURCES"
+                linux_name = "linux-%s.tar.xz" % self.output
+                tarball = "%s/%s" % (src_def_path, linux_name)
+                archive.extract(tarball, self.workdir)
+                os.chdir("%s/linux-%s/%s" % (self.workdir, self.output,
+                         self.testdir))
+            elif 'SuSE' in detected_distro.name:
+                if not smg.check_installed("kernel-source") and not\
+                        smg.install("kernel-source"):
+                    self.cancel(
+                        "Failed to install kernel-source for this test.")
+                if not os.path.exists("/usr/src/linux"):
+                    self.cancel("kernel source missing after install")
+                self.output = "/usr/src/linux"
+                os.chdir("%s/%s" % (self.output, self.testdir))
+        else:
+            location = self.params.get('location', default='https://github.c'
+                                       'om/torvalds/linux/archive/master.zip')
+            self.output = "master"
+
+            match = next(
+                (ext for ext in [".zip", ".tar"] if ext in location), None)
+            if match:
+                tarball = self.fetch_asset("kselftest%s" % match,
+                                           locations=[location], expire='1d')
+                archive.extract(tarball, self.workdir)
+            else:
+                git.get_repo(location, destination_dir=self.workdir)
+            os.chdir("%s/linux-%s/%s" % (self.workdir, self.output,
+                     self.testdir))
+
+    def test_ftrace_basic(self):
```

Comment:
instead of repeating each test with same steps each of the scenario can be make yaml configurable.